### PR TITLE
[Keyboard] add keyboard tnln95

### PR DESCRIPTION
**Pull request by NMHoang05**
* Added pom_keyboards folder
* Added keyboard tnln95

## Description

TNLN95 is the code name of my personal keyboard, which cosists of a 5 columns numpad and 65% layout.
Only basic layout is supported for now, I will add support for other layouts later.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/pom_keyboards/tnln95/config.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x544E
+#define PRODUCT_ID      0x3931
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    POM_keyboard
+#define PRODUCT         tnln95
+
+#define CONFIG_H
+
+#include "config_common.h"
```
Comment (zvecr) [inline keyboards/pom_keyboards/tnln95/config.h:27]:
```suggestion
#pragma once

#include "config_common.h"

/* USB Device descriptor parameter */
#define VENDOR_ID       0x544E
#define PRODUCT_ID      0x3931
#define DEVICE_VER      0x0001
#define MANUFACTURER    POM_keyboard
#define PRODUCT         tnln95
```

File: keyboards/pom_keyboards/tnln95/tnln95.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ */
+#pragma once
+
+#include "quantum.h"
+
+// This a shortcut to help you visually see your layout.
+// The first section contains all of the arguments
+// The second converts the arguments into a two-dimensional array
+
+#define LAYOUT_ALL(\
```
Comment (zvecr) [inline keyboards/pom_keyboards/tnln95/tnln95.h:24]:
```suggestion
#define LAYOUT_all(\
```

File: keyboards/pom_keyboards/tnln95/tnln95.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    { K30, K31, K32, K33, K34, K35, K36, K37, K38, K39 },\
+    { K40, K41, K42, K43, K44, K45, K46, K47, K48, K49 },\
+    { K50, K51, K52, K53, K54, K55, K56, KC_NO, K58, K59 },\
+    { K60, K61, K62, K63, K64, K65, K66, K67, K68, K69 },\
+    { K70, K71, K72, K73, K74, K75, KC_NO, K77, K78, K79 },\
+    { K80, K81, K82, K83, K84, K85, K86, K87, K88, K89 },\
+    { KC_NO, KC_NO, K92, K93, K94, K95, K96, K97, K98, K99 } \
+}
+
+#define LAYOUT_DEF_ANSI(\
```
Comment (zvecr) [inline keyboards/pom_keyboards/tnln95/tnln95.h:43]:
```suggestion
#define LAYOUT_ansi(\
```

File: keyboards/pom_keyboards/tnln95/keymaps/def_ansi/keymap.c [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+  [_DEF] = LAYOUT_DEF_ANSI(
+    KC_PMNS, KC_PAST, KC_PSLS, KC_BSPACE, KC_ESC,    KC_GRAVE, KC_1, KC_2, KC_3, KC_4, KC_5, KC_6, KC_7, KC_8, KC_9, KC_0, KC_MINUS, KC_EQUAL,     KC_BSPACE,    KC_HOME, \
+             KC_P7,   KC_P8,   KC_P9, KC_NUMLOCK,      KC_TAB, KC_Q, KC_W, KC_E, KC_R, KC_T, KC_Y, KC_U, KC_I, KC_O, KC_P, KC_LBRC, KC_RBRC,    KC_BSLASH,       KC_PGUP, \
+    KC_PPLS, KC_P4,   KC_P5,   KC_P6, CMD_COPY,          KC_CAPS, KC_A, KC_S, KC_D, KC_F, KC_G, KC_H, KC_J, KC_K, KC_L, KC_SCOLON, KC_QUOTE,       KC_ENTER,     KC_PGDN, \
+             KC_P1,   KC_P2,   KC_P3, CMD_PASTE,           KC_LSHIFT, KC_Z, KC_X, KC_C, KC_V, KC_B, KC_N, KC_M, KC_COMMA, KC_DOT, KC_SLASH,   KC_RSHIFT,   KC_UP, KC_END, \
+    KC_PENT, KC_PDOT,    KC_P0,      KC_COMMA ,       KC_LCTRL, KC_LGUI, KC_LALT,                KC_SPACE,              KC_RALT,   MO(_FNC),   KC_LEFT, KC_DOWN, KC_RIGHT \
+  ),
+  [_FNC] = LAYOUT_DEF_ANSI(
```
Comment (zvecr) [inline keyboards/pom_keyboards/tnln95/keymaps/def_ansi/keymap.c:32]:
```suggestion
  [_FNC] = LAYOUT_ansi
```

File: keyboards/pom_keyboards/tnln95/keymaps/def_ansi/keymap.c [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+#include QMK_KEYBOARD_H
+
+// Layer shorthand
+#define _DEF 0
+#define _FNC 1
+#define CMD_COPY LCTL(KC_INS)
+#define CMD_PASTE LSFT(KC_INS)
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+  [_DEF] = LAYOUT_DEF_ANSI(
```
Comment (zvecr) [inline keyboards/pom_keyboards/tnln95/keymaps/def_ansi/keymap.c:25]:
```suggestion
  [_DEF] = LAYOUT_ansi(
```

File: keyboards/pom_keyboards/tnln95/keymaps/def_ansi/keymap.c [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+// Layer shorthand
+#define _DEF 0
+#define _FNC 1
```
Comment (zvecr) [inline keyboards/pom_keyboards/tnln95/keymaps/def_ansi/keymap.c:20]:
```suggestion
// Defines names for use in layer keycodes and the keymap
enum layer_names {
    _DEF,
    _FNC
};

```

File: keyboards/pom_keyboards/tnln95/info.json [hunk 7]
```diff
@@ -0,0 +1,19 @@
+{
+    "keyboard_name": "tnln95",
+    "url": "http://www.keyboard-layout-editor.com/#/gists/45aee74c726e32a063130a715fc494d3",
+    "maintainer": "HoangNM",
+    "width": 21.25,
+    "height": 5,
+    "layouts": {
+        "LAYOUT_def_ansi": {
```
Comment (zvecr) [inline keyboards/pom_keyboards/tnln95/info.json:8]:
```suggestion
        "LAYOUT_ansi": {
```

File: keyboards/pom_keyboards/tnln95/rules.mk [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+MCU = atmega32u4
+
+# Bootloader selection
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   ATmega32A    bootloadHID
+#   ATmega328P   USBasp
```
Comment (zvecr) [inline keyboards/pom_keyboards/tnln95/rules.mk:11]:
```suggestion
```

File: keyboards/pom_keyboards/tnln95/rules.mk [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+BACKLIGHT_ENABLE = yes      # Enable keyboard backlight functionality on B7 by default
+RGBLIGHT_ENABLE = yes       # Enable RGB underglow
+MIDI_ENABLE = no            # MIDI controls
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+ENCODER_ENABLE = no         # enable the rotary encoder
+TAP_DANCE_ENABLE = no      # enable tap dancing
+EXTRAFLAGS += -flto
+
```
Comment (zvecr) [inline keyboards/pom_keyboards/tnln95/rules.mk:35]:
```suggestion
# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = yes      # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
LTO_ENABLE = yes
```

File: keyboards/pom_keyboards/tnln95/tnln95.c [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    
+    /* indicator for fn layer */
+    if (biton32(layer_state) == 1) {
+        PORTB |= (1<<3);
+    } else {
+        PORTB &= ~(1<<3);
+    }
+
+    led_set_user(usb_led);
+}
```
Comment (zvecr) [inline keyboards/pom_keyboards/tnln95/tnln95.c:59]:
```suggestion
void matrix_init_kb(void) {
    setPinOutput(B1);
    setPinOutput(B2);
    setPinOutput(B3);

    matrix_init_user();
}

bool led_update_kb(led_t led_state) {
    bool res = led_update_user(led_state);
    if(res) {
        writePin(B1, led_state.num_lock);
        writePin(B2, led_state.caps_lock);
    }
    return res;
}
```

File: keyboards/pom_keyboards/tnln95/keymaps/default/keymap.c [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+  ),
+  [_FNC] = LAYOUT_ansi(
+    _______, _______,   _______,   _______, _______,   _______,   KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_F9,   KC_F10,   KC_F11,   KC_F12,        KC_DEL,    KC_PSCR, \
+             _______,   _______,   _______, _______,      _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, RGB_TOG, RGB_RMOD, RGB_MOD,     KC_INS,      KC_VOLU, \
+    _______, _______,   _______,   _______, _______,          _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,         BL_TOGG,       KC_VOLD, \
+             _______,   _______,   _______, _______,           _______, _______, _______, _______, _______, _______, _______, _______, BL_DEC, BL_INC, BL_BRTG,         _______,        _______, KC_MUTE, \
+    _______, _______,       _______,        _______ ,       _______,   _______,   _______,                             _______,                      _______,     MO(_FNC),    _______, _______, _______ \
```
Comment (fauxpark) [inline keyboards/pom_keyboards/tnln95/keymaps/default/keymap.c:40]:
```suggestion
    KC_PMNS, KC_PAST, KC_PSLS, KC_BSPACE, KC_ESC,    KC_GRAVE, KC_1, KC_2, KC_3, KC_4, KC_5, KC_6, KC_7, KC_8, KC_9, KC_0, KC_MINUS, KC_EQUAL,     KC_BSPACE,    KC_HOME,
             KC_P7,   KC_P8,   KC_P9, KC_NUMLOCK,      KC_TAB, KC_Q, KC_W, KC_E, KC_R, KC_T, KC_Y, KC_U, KC_I, KC_O, KC_P, KC_LBRC, KC_RBRC,    KC_BSLASH,       KC_PGUP,
    KC_PPLS, KC_P4,   KC_P5,   KC_P6, CMD_COPY,          KC_CAPS, KC_A, KC_S, KC_D, KC_F, KC_G, KC_H, KC_J, KC_K, KC_L, KC_SCOLON, KC_QUOTE,       KC_ENTER,     KC_PGDN,
             KC_P1,   KC_P2,   KC_P3, CMD_PASTE,           KC_LSHIFT, KC_Z, KC_X, KC_C, KC_V, KC_B, KC_N, KC_M, KC_COMMA, KC_DOT, KC_SLASH,   KC_RSHIFT,   KC_UP, KC_END,
    KC_PENT, KC_PDOT,    KC_P0,      KC_COMMA ,       KC_LCTRL, KC_LGUI, KC_LALT,                KC_SPACE,              KC_RALT,   MO(_FNC),   KC_LEFT, KC_DOWN, KC_RIGHT
  ),
  [_FNC] = LAYOUT_ansi(
    _______, _______,   _______,   _______, _______,   _______,   KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_F9,   KC_F10,   KC_F11,   KC_F12,        KC_DEL,    KC_PSCR,
             _______,   _______,   _______, _______,      _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, RGB_TOG, RGB_RMOD, RGB_MOD,     KC_INS,      KC_VOLU,
    _______, _______,   _______,   _______, _______,          _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,         BL_TOGG,       KC_VOLD,
             _______,   _______,   _______, _______,           _______, _______, _______, _______, _______, _______, _______, _______, BL_DEC, BL_INC, BL_BRTG,         _______,        _______, KC_MUTE,
    _______, _______,       _______,        _______ ,       _______,   _______,   _______,                             _______,                      _______,     MO(_FNC),    _______, _______, _______
```

File: keyboards/pom_keyboards/tnln95/rules.mk [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+NKRO_ENABLE = yes            # USB Nkey Rollover
+BACKLIGHT_ENABLE = yes      # Enable keyboard backlight functionality on B7 by default
+RGBLIGHT_ENABLE = yes       # Enable RGB underglow
+MIDI_ENABLE = no            # MIDI controls
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+ENCODER_ENABLE = no         # enable the rotary encoder
+TAP_DANCE_ENABLE = no      # enable tap dancing
+EXTRAFLAGS += -flto
```
Comment (fauxpark) [inline keyboards/pom_keyboards/tnln95/rules.mk:27]:
```suggestion
BOOTMAGIC_ENABLE = lite      # Virtual DIP switch configuration
MOUSEKEY_ENABLE = no       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no        # Console for debug
COMMAND_ENABLE = no        # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes            # USB Nkey Rollover
BACKLIGHT_ENABLE = yes      # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = yes       # Enable RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
LTO_ENABLE = yes
```

File: keyboards/pom_keyboards/tnln95/readme.md [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+
+This is the default layout that comes flashed on every TNLN95.
+
+* Keyboard Maintainer: [Hoang Nguyen Minh](https://github.com/NMHoang05/)
+* Hardware Supported: TNLN95
+* Hardware Availability: (soon!)
+
+Make example for this keyboard (after setting up your build environment):
+
+    make pom_keyboards/tnln95:def_ansi
```
Comment (fauxpark) [inline keyboards/pom_keyboards/tnln95/readme.md:15]:
```suggestion
    make pom_keyboards/tnln95:default
```

File: keyboards/pom_keyboards/tnln95/config.h [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    #define RGBLED_NUM 14
+    #define RGBLIGHT_ANIMATIONS
+    #define RGBLIGHT_HUE_STEP 8
+    #define RGBLIGHT_SAT_STEP 8
+    #define RGBLIGHT_VAL_STEP 8
+    #define RGBLIGHT_SLEEP // RGB will turn off when PC is put to sleep
+    #define RGB_MATRIX_LED_PROCESS_LIMIT (DRIVER_LED_TOTAL + 4) / 5 // limits the number of LEDs to process in an animation per task run (increases keyboard responsiveness)
+    #define RGB_MATRIX_LED_FLUSH_LIMIT 16 // limits in milliseconds how frequently an animation will update the LEDs. 16 (16ms) is equivalent to limiting to 60fps (increases keyboard responsiveness)
+    #define RGB_MATRIX_MAXIMUM_BRIGHTNESS 225 // limits maximum brightness of LEDs to 200 out of 255. If not defined maximum brightness is set to 255
+    #define RGB_MATRIX_STARTUP_MODE RGB_MATRIX_SOLID_COLOR // Sets the default mode, if none has been set
```
Comment (drashna) [inline keyboards/pom_keyboards/tnln95/config.h:80]: These is the default value, and doesn't need to explicitly defined. 

```suggestion
//    #define RGB_MATRIX_LED_PROCESS_LIMIT (DRIVER_LED_TOTAL + 4) / 5 // limits the number of LEDs to process in an animation per task run (increases keyboard responsiveness)
//    #define RGB_MATRIX_LED_FLUSH_LIMIT 16 // limits in milliseconds how frequently an animation will update the LEDs. 16 (16ms) is equivalent to limiting to 60fps (increases keyboard responsiveness)
//    #define RGB_MATRIX_MAXIMUM_BRIGHTNESS 225 // limits maximum brightness of LEDs to 200 out of 255. If not defined maximum brightness is set to 255
//    #define RGB_MATRIX_STARTUP_MODE RGB_MATRIX_SOLID_COLOR // Sets the default mode, if none has been set
```

File: keyboards/pom_keyboards/tnln95/tnln95.c [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "tnln95.h"
+
+void matrix_init_kb(void) {
```
Comment (drashna) [inline keyboards/pom_keyboards/tnln95/tnln95.c:18]: I would prefer if pre-init was used here:
```suggestion
void keyboard_pre_init_kb(void) {
```

File: keyboards/pom_keyboards/tnln95/tnln95.c [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+    setPinOutput(B1);
+    setPinOutput(B2);
+    setPinOutput(B3);
+
+    matrix_init_user();
```
Comment (drashna) [inline keyboards/pom_keyboards/tnln95/tnln95.c:23]:
```suggestion
    keyboard_pre_init_user();
```

File: keyboards/pom_keyboards/tnln95/readme.md [hunk 17]
```diff
@@ -0,0 +1,20 @@
+# POM Keyboards - TNLN95
+
+![TNLN95 Layout Image](https://imgur.com/nIefJ0mp)
```
Comment (noroadsleft) [inline keyboards/pom_keyboards/tnln95/readme.md:3]: This link doesn't seem to work.

File: keyboards/pom_keyboards/tnln95/tnln95.c [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+bool led_update_kb(led_t led_state) {
+    bool res = led_update_user(led_state);
+    if(res) {
+        writePin(B1, led_state.num_lock);
+        writePin(B2, led_state.caps_lock);
+    }
+    return res;
+}
```
Comment (noroadsleft) [inline keyboards/pom_keyboards/tnln95/tnln95.c:33]: B3 is present in the init, but nothing happens with it after.

File: keyboards/pom_keyboards/tnln95/readme.md (hunk 17 above)
Comment (NMHoang05) [reply keyboards/pom_keyboards/tnln95/readme.md]: I fixed this (not sure how I can mis-copy the link :( )

File: keyboards/pom_keyboards/tnln95/tnln95.c (hunk 18 above)
Comment (NMHoang05) [reply keyboards/pom_keyboards/tnln95/tnln95.c]: I commented out B3 (and added a commented E2), I will add functionality for them in the future.

File: keyboards/pom_keyboards/tnln95/readme.md [hunk 19]
```diff
@@ -1,6 +1,6 @@
 # POM Keyboards - TNLN95
 
-![TNLN95 Layout Image](https://imgur.com/nIefJ0mp)
+![TNLN95 Layout Image](https://imgur.com/a/G4EfLta)
```
Comment (noroadsleft) [inline keyboards/pom_keyboards/tnln95/readme.md:3]: This is meant to be a direct image link.

```suggestion
![TNLN95 Layout Image](https://i.imgur.com/nIefJ0m.png)
```

**Comment by NMHoang05** [comment]
Hi, I saw "travis ci" build failed, 
I ran the command locally but did not get the error.
local travis build

can anyone help me resolve this?

**Comment by noroadsleft** [comment]
> can anyone help me resolve this?

You don't need to. Travis CI was testing every keyboard in the repo, and the failure is on a different keyboard. Everything is good here; thanks!

**Pull request merged**